src/common/rand/rand.h

```diff
@@ -66,6 +66,16 @@ OQS_API void OQS_randombytes(uint8_t *random_array, size_t bytes_to_read);
  */
 OQS_API void OQS_randombytes_nist_kat_init_256bit(const uint8_t *entropy_input, const uint8_t *personalization_string);
 
+/**
+ * Saves the state of the NIST DRBG, allowing it to be recovered later.
+ */
+OQS_API void OQS_randombytes_nist_kat_save_state(void);
```

Comment:
I'm absolutely not sure I like this API design: It doesn't seem idempotent (what if someone calls this twice, before calling `restore`?) and creating state in the library. What about returning here an opaque structure/pointer and taking this as parameter in the restore below instead?

Reply:
One could do that, but I believe it would require either (1) "exporting" the `AES256_CTR_DRBG_ctx` structure so that it can be used by the test programs or (2) duplicating the definition of said structure in each test program that requires it (currently only `kat_kem` and `kat_sig`). Both of those are OK, I guess, but I don't think that the required "carving up" of the NIST DRBG code would be worth it: I doubt there's any use case that will require successive "save" calls without a "restore", and there is already state stored to enable the NIST DRBG.

I'd personally rather not even make the "save" and "restore" functions part of the API, but it seems to be the only clean(ish) way to make them accessible to our test programs...

Reply:
> I'd personally rather not even make the "save" and "restore" functions part of the API, 

Then it's two of us...

> but it seems to be the only clean(ish) way to make them accessible to our test programs...

...so then we probably do something wrong: Things only to be used by `liboqs` testing ought to have an internal API. But it seems we only have one `liboqs` API, namely a very public one. My understanding of public APIs is that they need to be maintained (and maintainable). Adding something to such public API should be carefully considered. I'd strongly suggest moving features we only need internally into a truly internal API, maybe something like "test_helpers.h/c" in the "tests" directory?

Reply:
How about this approach:

- a `NIST_KAT_DRBG_STATE` struct
- `void nist_kat_get_state(NIST_KAT_DRBG_STATE *)`
- `void nist_kat_set_state(const NIST_KAT_DRBG_STATE *)`

The "save" and "restore" functionality can be implemented using these functions in the new "test_helpers" files.

This seems to me to be the least disruptive option: the NIST DRBG continues to function as it does currently (library state), and `get_state` and `set_state` provide the minimal functionality required to implement save and restore. It's also clear what they do: no issues with idempotence etc.

I realize this sounds very much like option (1) which I objected to above, but using the get/set paradigm instead of save/restore makes it more palatable imo.

Reply:
Is it necessary to expose the struct? Couldn't we make do with a generic `void*`? Otherwise the proposal looks good (even though I'd have preferred an option to completely leave the API untouched -- but given we already expose the NIST RNG (incl. state) I also don't see a way around this :(

Reply:
> Is it necessary to expose the struct? Couldn't we make do with a generic `void*`?

Sure, that should be fine. I'll get to work on it.